middleman_protocol/middleman_protocol/helpers.py

```diff
@@ -0,0 +1,81 @@
+import socket
+
+from mypy.types import Iterable
+
+from middleman_protocol.constants import ESCAPE_CHARACTER
+from middleman_protocol.constants import ESCAPE_SEQUENCES
+from middleman_protocol.constants import FRAME_PAYLOAD_LENGTH_STARTING_BYTE
+from middleman_protocol.constants import FRAME_PAYLOAD_STARTING_BYTE
+from middleman_protocol.constants import FRAME_SEPARATOR
+from middleman_protocol.constants import FRAME_SEPARATOR_LENGTH
+from middleman_protocol.constants import RECEIVE_BYTES_PER_LOOP
+from middleman_protocol.message import AbstractMiddlemanMessage
+
+
+def handle_message_receive(connection: socket.socket, encoding_expected: bool=False) -> Iterable[bytes]:
+    """
+    Helper generator for receiving serialized Middleman protocol messages.
+
+    Yields received bytes.
+    """
+    assert isinstance(connection, socket.socket)
+    assert isinstance(encoding_expected, bool)
+
+    received_data = b''
+    message_with_payload_length = None
+    # Used to mark how many bytes was received since last decoding, to avoid decoding twice same part of data.
+    decoding_buffer = 0
+
+    while True:
+        received_data += connection.recv(RECEIVE_BYTES_PER_LOOP)
+
+        # If encoding is expected to be used by data sender,
+        # and there is encoded sequence in last received data,
+        # decode it and set encoding buffer.
+        if (
+            encoding_expected is True and (
+                ESCAPE_SEQUENCES[FRAME_SEPARATOR] in received_data[decoding_buffer:] or
+                ESCAPE_SEQUENCES[ESCAPE_CHARACTER] in received_data[decoding_buffer:]
+            )
+        ):
+            received_data = (
+                received_data[:decoding_buffer] +
+                AbstractMiddlemanMessage.escape_decode_raw_message(received_data[decoding_buffer:])
+            )
+            decoding_buffer = len(received_data)
+
+        # If separator is met, reset received data and restart.
+        if (
+            len(received_data) >= FRAME_SEPARATOR_LENGTH and
+            received_data[-FRAME_SEPARATOR_LENGTH:] == FRAME_SEPARATOR
+        ):
```

Comment:
Ok I have changed it just like that and all tests (along with those newly added for edge cases with escaping) still passes .. so it should be ok with that change. Thinking about this, I have also removed what I have added in 8ed6143cf76d006093c6557c4cacfb9a56cf778d (previously tests didn't pass without it), so maybe there was just enough to switch those two blocks.

Reply:
So maybe you're missing a test. Consider this case:

|                          |               |
|--------------------------|---------------|
| Separator                | `66 66 66`
| Escape character         | `33 33 33`
| Escaped separator        | `33 33 33 00`
| Escaped escape character | `33 33 33 01`

Input:
```
22 33 33 33 00 44 BB 66 66 66
```

The frame after unescaping and stripping separators should be
```
22 66 66 66 44 BB
```

But if you unescape first you would see:
```
22 66 66 66 44 BB 66 66 66
```

and then you would strip separators and see two messages instead of one:
```
22
44 BB
```

Reply:
But it won't happen because current flow in handler is different. It the way you described it is:
1. Unescaping the whole received data (with separatator).
2. Splitting unescaped data into smaller pieces (supposedly frames).

But the handler doesn't work this way. It doesn't ever return message with separator (the original one, not the one which is result of unescaping). And after returning the data there is no extra splitting into smaller messages. 

Maybe this is worth describing in more detailed way in handler docstring, but the situation which you have described won't happen if you use handler as expected.

Reply:
https://github.com/golemfactory/concent/pull/673#discussion_r206953885 shows why it's natural to treat these things as separate. Removal of separator naturally fits into a generator that's a layer below the unescaping and is completely independent.